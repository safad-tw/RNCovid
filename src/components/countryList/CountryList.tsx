import * as React from 'react';
import { useState } from 'react';
import {
  Text,
  View,
  StyleSheet,
  FlatList, TextInput
} from 'react-native';

import Filters from './Filters'
import LanguagesStrings from './../../localization/LanguagesStrings';


function CountryList(props: any) {

  const data = props.route.params.data.Countries
  const [query, setQuery] = useState('');
  const [filterData, setFilterData] = useState(data);
  const [filters] = useState([
    { label: LanguagesStrings.recovered },
    { label: LanguagesStrings.deaths },
    { label: LanguagesStrings.active}
  ]);
  const [selected, setSelected] = useState(filters[0]);

  const filterByDeaths = () => {
    const sortedCountries = data.sort(function (a: any, b: any) {
      return parseFloat(b.TotalDeaths) - parseFloat(a.TotalDeaths);
    });
    setFilterData(sortedCountries)
  }

  const filterByActiveCases = () => {
    const sortedCountries = data.sort(function (a: any, b: any) {
      return parseFloat(b.TotalConfirmed) - parseFloat(a.TotalConfirmed);
    });
    setFilterData(sortedCountries)
  }

  const filterByRecovered = () => {
    const sortedCountries = data.sort(function (a: any, b: any) {
      return parseFloat(b.TotalRecovered) - parseFloat(a.TotalRecovered);
    });
    setFilterData(sortedCountries)
  }

  const callback = (id: number) => {
    switch(id) {
      case 0:
        setSelected(filters[0])
        filterByRecovered()
        break;   
      case 1:
        setSelected(filters[1])
        filterByDeaths()
        break;
      case 2:
        setSelected(filters[2])
        filterByActiveCases()
         break;
      default:
        break;       
    }
  }

  const handleSearch = (text: string) => {
    const formattedQuery = text.toLowerCase();
    const filteredData = data.filter(function (a: any) {
      return a.Country.toLowerCase().includes(formattedQuery)
    });
    setFilterData(filteredData);
    setQuery(text);
  };

  return (
    <View style={styles.container}>
      <View
        style={styles.searchBarContainer}
      >
        <TextInput
          autoCapitalize="none"
          autoCorrect={false}
          clearButtonMode="always"
          value={query}
          onChangeText={queryText => handleSearch(queryText)}
          placeholder={LanguagesStrings.searchByCountry}
          style={{ backgroundColor: '#fff', paddingHorizontal: 20 }}
        />
      </View>
      <View style={{ flexDirection: 'row', margin: 5 }}>
        <Filters data={filters} onValueChange={(id) => callback(id)} />
      </View>
      <FlatList
        data={filterData}
        keyExtractor={item => item.CountryCode}
        renderItem={({ item }) => (
          <View style={styles.listItem}>
            <Text style={styles.listItemText}>{item.Country}</Text>
            <Text style={styles.listItemCaseText}>Total Confirmed: {item.TotalConfirmed}</Text>
            <Text style={styles.listItemCaseText}>TotalDeaths: {item.TotalDeaths}</Text>
            <Text style={styles.listItemCaseText}>NewConfirmed: {item.NewConfirmed}</Text>
            <Text style={styles.listItemCaseText}>NewDeaths: {item.NewDeaths}</Text>
            <Text style={styles.listItemCaseText}>NewRecovered: {item.NewRecovered}</Text>
            <Text style={styles.listItemCaseText}>TotalRecovered: {item.TotalRecovered}</Text>
          </View>
        )}
      />
    </View>
  )

}

export default CountryList;

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#f8f8f8',
    width: '100%',
  },
  listItem: {
    margin: 10,
    width: '95%',
    backgroundColor: '#F5FCFF',
    borderRadius: 5,
    shadowColor: 'gray',
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 0.8,
    shadowRadius: 2,
    elevation: 5,
  },
  listItemText: {
    margin: 5,
    fontSize: 18,
    fontFamily: "Avenir-black",
  },
  listItemCaseText: {
    margin: 5,
    fontSize: 16,
    fontFamily: "Avenir-medium"
  },
  searchBarContainer: {
    backgroundColor: '#fff',
    padding: 10,
    margin: 10,
    borderRadius: 20,
    width: '95%',
    fontSize: 16,
    fontFamily: "Avenir-medium"
  },
  fliter_container: {
    flexDirection: 'row',
    backgroundColor: '#ecf0f1',
    padding: 8,
  },
  filterButton: {
    borderRadius: 15, 
    padding: 8, 
    paddingLeft: 24, 
    paddingRight: 24,
    marginRight: 8
  },
});