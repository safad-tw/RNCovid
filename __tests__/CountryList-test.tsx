import * as React from "react";
import CountryList from "../src/components/countryList/CountryList";
import { render } from '@testing-library/react-native';
import '@testing-library/jest-dom'


const data: any = [
    {
      ID: "d81abd15-3c81-4c98-82e7-6042959daa6d",
      Country: "Afghanistan",
      CountryCode: "AF",
      Slug: "afghanistan",
      NewConfirmed: 0,
      TotalConfirmed: 155093,
      NewDeaths: 0,
      TotalDeaths: 7201,
      NewRecovered: 0,
      TotalRecovered: 0,
      Date: "2021-09-29T12:04:59.613Z",
      Premium: {}
    },
    {
      ID: "39714019-ce94-461b-b81a-1236f961c02a",
      Country: "Albania",
      CountryCode: "AL",
      Slug: "albania",
      NewConfirmed: 0,
      TotalConfirmed: 168782,
      NewDeaths: 0,
      TotalDeaths: 2668,
      NewRecovered: 0,
      TotalRecovered: 0,
      Date: "2021-09-29T12:04:59.613Z",
      Premium: {}
    },
    {
      ID: "729b2b5d-848c-4597-98b2-d35a90045c80",
      Country: "Algeria",
      CountryCode: "DZ",
      Slug: "algeria",
      NewConfirmed: 0,
      TotalConfirmed: 203045,
      NewDeaths: 0,
      TotalDeaths: 5797,
      NewRecovered: 0,
      TotalRecovered: 0,
      Date: "2021-09-29T12:04:59.613Z",
      Premium: {}
    },
    {
      ID: "da4c9b7a-5f18-4105-a349-0af456ebc00b",
      Country: "Andorra",
      CountryCode: "AD",
      Slug: "andorra",
      NewConfirmed: 0,
      TotalConfirmed: 15192,
      NewDeaths: 0,
      TotalDeaths: 130,
      NewRecovered: 0,
      TotalRecovered: 0,
      Date: "2021-09-29T12:04:59.613Z",
      Premium: {}
    },
    {
      ID: "e0737d69-f4ee-417d-b781-83ceb157efd9",
      Country: "Angola",
      CountryCode: "AO",
      Slug: "angola",
      NewConfirmed: 0,
      TotalConfirmed: 56040,
      NewDeaths: 0,
      TotalDeaths: 1526,
      NewRecovered: 0,
      TotalRecovered: 0,
      Date: "2021-09-29T12:04:59.613Z",
      Premium: {}
    }];
const createTestProps = (props: Object) => ({
    route:{
        params:{
            data:data
        }
    },
    navigation: {
      navigate: jest.fn()
    },
    ...props
  });

  describe("Country list", () => {
    describe("rendering", () => {
      let wrapper: any;
      let props: any;   
      beforeEach(() => {
        props = createTestProps({});
        wrapper = render(<CountryList {...props} />);
      });
  
      it("should render a <View />", () => {
        expect(wrapper).not.toBeNull()   // SUCCESS
      });

      it('Renders snapshot as expected', () => {
        expect(wrapper).toMatchSnapshot();
      });

    
    });
  });



  