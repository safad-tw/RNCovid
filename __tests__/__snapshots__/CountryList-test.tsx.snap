// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Country list rendering Renders snapshot as expected 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#f8f8f8",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 20,
        "fontFamily": "Avenir-medium",
        "fontSize": 16,
        "margin": 10,
        "padding": 10,
        "width": "95%",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      clearButtonMode="always"
      onChangeText={[Function]}
      placeholder="Search by country name"
      style={
        Object {
          "backgroundColor": "#fff",
          "paddingHorizontal": 20,
        }
      }
      value=""
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "margin": 5,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "black",
            "borderRadius": 20,
            "borderWidth": 2,
            "margin": 5,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
            }
          }
        >
          Recovered
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "black",
            "borderRadius": 20,
            "borderWidth": 2,
            "margin": 5,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
            }
          }
        >
          Deaths
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "black",
            "borderRadius": 20,
            "borderWidth": 2,
            "margin": 5,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
            }
          }
        >
          Active
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</View>
`;

exports[`LoadingScreen rendering Renders snapshot as expected 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#f8f8f8",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 20,
        "fontFamily": "Avenir-medium",
        "fontSize": 16,
        "margin": 10,
        "padding": 10,
        "width": "95%",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      clearButtonMode="always"
      onChangeText={[Function]}
      placeholder="Search by country name"
      style={
        Object {
          "backgroundColor": "#fff",
          "paddingHorizontal": 20,
        }
      }
      value=""
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "margin": 5,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "black",
            "borderRadius": 20,
            "borderWidth": 2,
            "margin": 5,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
            }
          }
        >
          Recovered
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "black",
            "borderRadius": 20,
            "borderWidth": 2,
            "margin": 5,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
            }
          }
        >
          Deaths
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "black",
            "borderRadius": 20,
            "borderWidth": 2,
            "margin": 5,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
            }
          }
        >
          Active
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</View>
`;
